### PR TITLE
[wasm][build] Build stdlib and test IRGen for WebAssembly

### PR DESCRIPTION
**Pull request by kateinoigakukun**
This patch adds new build products to build and test the Swift stdlib for WebAssembly.
This adds WebAssembly specific stdlib product instead of adding the new target in `stdlib-deployment-targets` because unlike darwin platforms Wasm target has quite different stdlib configuration from host target and there is no way to specify different stdlib configuration for each target in `stdlib-deployment-targets` for now.

This patch adds 1 new dependency `wasi-libc` and 3 new build products:
1. `WASILibc` - wasi-libc is a libc implementation for WebAssembly System Interface (WASI). This product is required to build the compiler runtimes and Swift stdlib.
2. `WasmLLVMRuntimeLibs` - This product builds `./runtimes` directory in llvm-project, which contains compiler-rt, libcxx and libcxxabi. This product is required to build Swift stdlib.
3. `WasmStdlib` - This product builds standalone Swift stdlib for WebAssembly. (Similar to `MinimalStdlib` product, but configured for WebAssembly target)

Currently `WasmStdlib` runs test suite under `test/IRGen` to verify the semi-target specific code in IRGen works and the generated IR can be processed by LLVM WebAssembly backend. Rest of test suites including executable tests are now skipped, but will be enabled after we set up a Wasm runtime in the CI infrastructure and calling convention mismatch issues will be resolved. 

Resolves rdar://116007434 and rdar://116523979.

### CODE REVIEW

**Comment by kateinoigakukun** [comment]
@swift-ci Please smoke test

**Comment by kateinoigakukun** [comment]
preset= buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

**Comment by kateinoigakukun** [comment]
@swift-ci Please smoke test

**Comment by kateinoigakukun** [comment]
preset=buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

**Comment by kateinoigakukun** [comment]
The stdlib compilation crash will be resolved after rebranch will be merged

**Comment by kateinoigakukun** [comment]
I confirmed it passes with the llvm fix included in the rebranch https://github.com/apple/swift/pull/69239

File: utils/update_checkout/update-checkout-config.json [hunk 1]
```diff
@@ -131,7 +133,8 @@
                 "swift-markdown": "main",
                 "swift-nio": "2.31.2",
                 "swift-nio-ssl": "2.15.0",
-                "swift-experimental-string-processing": "swift/main"
+                "swift-experimental-string-processing": "swift/main",
+                "wasi-libc": "wasi-sdk-20"
```
Comment (shahmishal) [inline utils/update_checkout/update-checkout-config.json:137]: Also add this to `rebranch` and `next` schemes.

File: test/IRGen/async_dynamic_replacement.swift [hunk 2]
```diff
@@ -1,6 +1,8 @@
 // RUN: %target-swift-frontend %s -emit-ir -disable-availability-checking -disable-objc-interop | %FileCheck %s
 
 // REQUIRES: concurrency
+// LLVM does not support swifttailcc for WebAssembly target for now (rdar://116753879)
```
Comment (shahmishal) [inline test/IRGen/async_dynamic_replacement.swift:4]: Can we create an issue on GitHub too?

**Comment by kateinoigakukun** [comment]
preset=buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

**Comment by kateinoigakukun** [comment]
@swift-ci Please smoke test

**Comment by kateinoigakukun** [comment]
preset=buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

**Comment by kateinoigakukun** [comment]
@swift-ci Please smoke test

**Comment by kateinoigakukun** [comment]
Added a few test fix commits that are found after rebranch 🩹

**Comment by kateinoigakukun** [comment]
preset=buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

**Comment by kateinoigakukun** [comment]
@swift-ci Please smoke test

**Comment by kateinoigakukun** [comment]
preset=buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

**Comment by kateinoigakukun** [comment]
@swift-ci Please smoke test

File: stdlib/public/Platform/wasi-libc.modulemap [hunk 3]
```diff
@@ -19,20 +19,12 @@ module SwiftWASILibc [system] {
   header "ctype.h"
   header "errno.h"
   header "fenv.h"
-  header "float.h"
```
Comment (MaxDesiatov) [inline stdlib/public/Platform/wasi-libc.modulemap:22]: Are these no longer needed or are they transitively included by other headers?

Comment (kateinoigakukun) [reply stdlib/public/Platform/wasi-libc.modulemap:22]: They are now part of `_Builtin_XXX` modules provided by clang https://reviews.llvm.org/D159064

**Review by MaxDesiatov — APPROVED** [review]
🤩 Super excited to see this, thank you!

**Review by shahmishal — APPROVED** [review]

**Pull request merged**